### PR TITLE
Implement new indicator McGinleyDynamic

### LINKED ISSUE #8127: Implement McGinley Dynamic
#### Expected Behavior
Supports McGinley Dynamic (see [talipp implementation](https://github.com/nardew/talipp/blob/main/talipp/indicators/McGinleyDynamic.py)).

Refer to [LEAN Engine > Contributions > Indicators](https://www.quantconnect.com/docs/v2/lean-engine/contributions/indicators).

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by femtotrader**
#### Description

Implement new indicator - McGinley Dynamic

#### Related Issue
Closes #8127 

#### Motivation and Context

McGinley Dynamic wasn't implemented in QuantConnect

#### Requires Documentation Change

Doc for this indicator should be created

#### How Has This Been Tested?

comparison with talipp computation of ZLEMA(5) of SPY close prices

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Hey @femtotrader!
Please take https://github.com/QuantConnect/Lean/pull/8148 as a reference of what should be changed in the other PRs 🙏, same review applies for all. Needs a rebase btw

**Comment by femtotrader** [comment]
Thanks @Martin-Molinero 
I fixed the merge conflict but I don't really see what is blocking here.
Some guidance will be nice on this one also (or simply take the lead on the PR)

File: Indicators/McGinleyDynamic.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// Represents the McGinley Dynamic (MGD)
```
Comment (Martin-Molinero) [inline Indicators/McGinleyDynamic.cs:21]: Please provide a small description of the `McGinleyDynamic` is 👍

File: Indicators/McGinleyDynamic.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    /// </summary>
+    public class McGinleyDynamic : WindowIndicator<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        /// <summary>
+        /// A rolling sum for computing the average for the given period
+        /// </summary>
+        private IndicatorBase<IndicatorDataPoint> _rollingSum { get; }
```
Comment (Martin-Molinero) [inline Indicators/McGinleyDynamic.cs:28]: For privates can directly use readonly field in this case 👍

File: Indicators/McGinleyDynamic.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        public override bool IsReady => Samples >= _period;
+
+        /// <summary>
+        /// Resets this indicator to its initial state
+        /// </summary>
+        public override void Reset()
+        {
+            _rollingSum.Reset();
+            base.Reset();
+        }
```
Comment (Martin-Molinero) [inline Indicators/McGinleyDynamic.cs:44]: nit - Please push after constructor

File: Indicators/McGinleyDynamic.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        /// <summary>
+        /// Computes the next value for this indicator from the given state.
+        /// </summary>
+        /// <param name="window">The window of data held in this indicator</param>
+        /// <param name="input">The input value to this indicator on this time step</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IReadOnlyWindow<IndicatorDataPoint> window, IndicatorDataPoint input)
+        {
+            _rollingSum.Update(input.Time, input.Value);
+            if (Samples < _period) {
```
Comment (Martin-Molinero) [inline Indicators/McGinleyDynamic.cs:76]: - Can directly use the output of `_rollingSum.Update` to know if it's ready or not?
- Coding style is off, please expand `if` brackets

File: Indicators/McGinleyDynamic.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                if (Current.Value == 0)
+                    return Current.Value;
+                return Current.Value + (input.Value - Current.Value) / (_period * (decimal)Math.Pow((double)(input.Value / Current.Value), 4.0));
+            }
+        }
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod => Period;
```
Comment (Martin-Molinero) [inline Indicators/McGinleyDynamic.cs:90]: Missing the `override` key word, given the base is virtual, should be moved further up to follow coding style

File: Indicators/McGinleyDynamic.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        {
+            _rollingSum.Update(input.Time, input.Value);
+            if (Samples < _period) {
+                return 0;
+            } else if (Samples == _period) {
+                return _rollingSum.Current.Value / _period;
+            } else {
+                if (Current.Value == 0)
+                    return Current.Value;
+                return Current.Value + (input.Value - Current.Value) / (_period * (decimal)Math.Pow((double)(input.Value / Current.Value), 4.0));
```
Comment (Martin-Molinero) [inline Indicators/McGinleyDynamic.cs:83]: We should probably add a protection against `/ 0` or infinity explosions

File: Tests/Indicators/McGinleyDynamicTests.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        protected override string TestFileName => "spy_with_McGinleyDynamic.csv";
+
+        protected override string TestColumnName => "McGinleyDynamic14";
+
+        protected override Action<IndicatorBase<IndicatorDataPoint>, double> Assertion
+        {
+            get { return (indicator, expected) => Assert.AreEqual(expected, (double)indicator.Current.Value, 1e-2); }
+        }
```
Comment (Martin-Molinero) [inline Tests/Indicators/McGinleyDynamicTests.cs:37]: Why add this override? The base has an assertion with a higher precision even

File: Tests/Indicators/McGinleyDynamicTests.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                indicator.Update(data);
+            }
+            Assert.IsTrue(indicator.IsReady);
+
+            indicator.Reset();
+
+            TestHelper.AssertIndicatorIsInDefaultState(indicator);
+            indicator.Update(DateTime.UtcNow, 2.0m);
+            indicator.Update(DateTime.UtcNow, 2.0m);
+            indicator.Update(DateTime.UtcNow, 2.0m);
```
Comment (Martin-Molinero) [inline Tests/Indicators/McGinleyDynamicTests.cs:67]: Please replace `DateTime.UtcNow`

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Hey @femtotrader! 
Thank you! 👍  
Posting some comments, basically the same as the ones in https://github.com/QuantConnect/Lean/pull/8148 and the commit I pushed there, please take a look, they apply to all indicator PRs

File: Algorithm/QCAlgorithm.Indicators.cs [hunk 9]
```diff
@@ -2514,6 +2514,22 @@ public McClellanSummationIndex MSI(Symbol[] symbols, int fastPeriod = 19, int sl
...
         /// <summary>
+        /// Creates a new McGinley Dynamic indicator
+        /// </summary>
+        /// <param name="symbol">The symbol whose McGinley Dynamic indicator value we want</param>
+        /// <param name="period">The period of the McGinley Dynamic indicator</param>
+        /// <param name="resolution">The resolution</param>
+        /// <param name="selector">Selects a value from the BaseData to send into the indicator, if null defaults to casting the input value to a TradeBar</param>
+        /// <returns>The McGinley Dynamic indicator for the requested symbol over the specified period</returns>
+        [DocumentationAttribute(Indicators)]
+        public McGinleyDynamic MGD(Symbol symbol, int period, Resolution? resolution = null, Func<IBaseData, decimal> selector = null)
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Algorithm/QCAlgorithm.Indicators.cs:2525]: ok was thinking that McGinley should be ranked after McClellan but it's function name that matter

File: Indicators/McGinleyDynamic.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        /// A rolling sum for computing the average for the given period
+        /// </summary>
+        private IndicatorBase<IndicatorDataPoint> _rollingSum { get; }
+
+        private readonly int _period;
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady => Samples >= _period;
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/McGinleyDynamic.cs:35]: Hum... IsReady() returns a boolean (2 values)... we have 3 code paths.

I can change

    Samples < _period

for

    !IsReady

but can't get rid of 

    Samples == _period

The problem here is we should have a 3 states variable (an enum?) to describe what is indicator state

- WarmingUp
- Warm
- Ready

I haven't see anything like that in your code base but it should probably be considered

File: Indicators/McGinleyDynamic.cs (hunk 1 above)
Comment (femtotrader) [reply Indicators/McGinleyDynamic.cs:21]: Done. Given that MGD is also a moving average I also added it to MovingAverageType.cs and MovingAverageTypeExtensions.cs

File: Indicators/McGinleyDynamic.cs (hunk 6 above)
Comment (Martin-Molinero) [reply Indicators/McGinleyDynamic.cs:88]: nit - missing `{}` too
Should rebase after https://github.com/QuantConnect/Lean/pull/8177 👍

**Comment by femtotrader** [comment]
Oups... I did a mistake when trying to rebase.

**Pull request closed without merging**